File: src/dpl_pipeline/aggregate.py
"""Aggregate provision-level matches into Impact/Bill records.

One output record per (bill, impact category) with at least one qualifying
provision; a bill with no qualifying provisions gets a single `neutral`
record. A bill is flagged `needs_review` when any single provision matched
rules in more than one category (e.g. the same section fired both a
restrictive and an expansive rule) -- that usually means the section is
genuinely mixed or a rule needs tightening, and a human should look.
"""

from __future__ import annotations

from collections import defaultdict

from .models import CATEGORIES, NEUTRAL, BillResult, ImpactRecord

_MAX_EVIDENCE = 8


def aggregate(result: BillResult) -> list[ImpactRecord]:
    bill = result.bill

    if result.status != "ok":
        return [
            ImpactRecord(
                bill_number=bill.bill_number,
                state=bill.state,
                category="unprocessed",
                policy_areas=(),
                rule_ids=(),
                n_provisions=0,
                evidence=(),
                source_used=result.source_used,
                text_sha256=result.text_sha256,
                status=result.status,
                needs_review=True,
            )
        ]

    # Which provisions matched multiple categories?
    cats_by_provision: dict[str, set[str]] = defaultdict(set)
    for m in result.matches:
        cats_by_provision[m.provision_id].add(m.category)
    mixed_provisions = {p for p, cats in cats_by_provision.items() if len(cats) > 1}

    by_category: dict[str, list] = defaultdict(list)
    for m in result.matches:
        by_category[m.category].append(m)

    records: list[ImpactRecord] = []
    for category in CATEGORIES:
        cat_matches = by_category.get(category)
        if not cat_matches:
            continue
        cat_matches.sort(key=lambda m: (m.provision_id, m.rule_id))
        evidence = tuple(
            f"{m.provision_id}|{m.rule_id}|{m.excerpt}" for m in cat_matches[:_MAX_EVIDENCE]
        )
        records.append(
            ImpactRecord(
                bill_number=bill.bill_number,
                state=bill.state,
                category=category,
                policy_areas=tuple(sorted({m.policy_area for m in cat_matches})),
                rule_ids=tuple(sorted({m.rule_id for m in cat_matches})),
                n_provisions=len({m.provision_id for m in cat_matches}),
                evidence=evidence,
                source_used=result.source_used,
                text_sha256=result.text_sha256,
                status=result.status,
                needs_review=any(m.provision_id in mixed_provisions for m in cat_matches),
            )
        )

    if not records:
        records.append(
            ImpactRecord(
                bill_number=bill.bill_number,
                state=bill.state,
                category=NEUTRAL,
                policy_areas=(),
                rule_ids=(),
                n_provisions=0,
                evidence=(),
                source_used=result.source_used,
                text_sha256=result.text_sha256,
                status=result.status,
                needs_review=False,
            )
        )
    return records

File: lib/aggregate.ts
/**
 * Aggregate provision-level matches into Impact/Bill records.
 *
 * One record per (bill, impact category) with at least one qualifying
 * provision; a bill with no qualifying provisions gets a single `neutral`
 * record. A record is flagged `needs_review` when any contributing provision
 * matched rules in more than one category — that usually means the section is
 * genuinely mixed or a rule needs tightening, and a human should look.
 *
 * A bill that could not be fetched/extracted yields one `unprocessed` record
 * (never silently `neutral`).
 */

import { CATEGORIES, NEUTRAL, billState, type BillResult, type ImpactRecord, type RuleMatch } from "./types";

const MAX_EVIDENCE = 8;

export function aggregate(result: BillResult): ImpactRecord[] {
  const bill = result.bill;
  const base = {
    bill_number: bill.bill_number,
    state: billState(bill),
    source_used: result.sourceUsed,
    text_sha256: result.textSha256,
    status: result.status,
  };

  if (result.status !== "ok") {
    return [
      {
        ...base,
        category: "unprocessed",
        policy_areas: [],
        rule_ids: [],
        n_provisions: 0,
        evidence: [],
        needs_review: true,
      },
    ];
  }

  const catsByProvision = new Map<string, Set<string>>();
  for (const m of result.matches) {
    if (!catsByProvision.has(m.provisionId)) catsByProvision.set(m.provisionId, new Set());
    catsByProvision.get(m.provisionId)!.add(m.category);
  }
  const mixedProvisions = new Set(
    [...catsByProvision.entries()].filter(([, cats]) => cats.size > 1).map(([p]) => p),
  );

  const byCategory = new Map<string, RuleMatch[]>();
  for (const m of result.matches) {
    if (!byCategory.has(m.category)) byCategory.set(m.category, []);
    byCategory.get(m.category)!.push(m);
  }

  const records: ImpactRecord[] = [];
  for (const category of CATEGORIES) {
    const catMatches = byCategory.get(category);
    if (!catMatches?.length) continue;
    catMatches.sort((a, b) =>
      a.provisionId === b.provisionId
        ? a.ruleId.localeCompare(b.ruleId)
        : a.provisionId.localeCompare(b.provisionId),
    );
    records.push({
      ...base,
      category,
      policy_areas: [...new Set(catMatches.map((m) => m.policyArea))].sort(),
      rule_ids: [...new Set(catMatches.map((m) => m.ruleId))].sort(),
      n_provisions: new Set(catMatches.map((m) => m.provisionId)).size,
      evidence: catMatches.slice(0, MAX_EVIDENCE).map((m) => ({
        provision_id: m.provisionId,
        rule_id: m.ruleId,
        excerpt: m.excerpt,
      })),
      needs_review: catMatches.some((m) => mixedProvisions.has(m.provisionId)),
    });
  }

  if (records.length === 0) {
    records.push({
      ...base,
      category: NEUTRAL,
      policy_areas: [],
      rule_ids: [],
      n_provisions: 0,
      evidence: [],
      needs_review: false,
    });
  }
  return records;
}
